// @cocreate/docs/src/socket_process.js
const CoCreate.socket = require('./CoCreate-socket')

module.exports.CoCreate.socketInit = function (socket) {
	CoCreate.socket.create({
		namespace: socket.config.organization_Id,
		room: null,
		host: socket.host
	})
	CoCreate.socket.setGlobalScope(socket.config.organization_Id);
}

module.exports.CoCreateUpdateDocument = function (info, config) {
	if (info === null) return;
	
	let commonParams = {
      "apiKey":           config.apiKey,
      "securityKey":      config.securityKey,
      "organization_id":  config.organization_Id,
	}
	
	let request_data = {...commonParams};
	
	if (!info.collection || !info.document_id ) return;
	
	request_data['set'] = info.data;
	request_data['collection'] = info.collection
	request_data['document_id'] = info.document_id
	request_data['metadata'] = info['metadata']
	request_data['upsert'] = true;
	request_data['broadcast'] = false;

	CoCreate.socket.send('updateDocument', request_data, '');	
}

module.exports.CoCreateCreateDocument = function (info, config) {
	if (info === null) return;
	
	let commonParams = {
      "apiKey":           config.apiKey,
      "securityKey":      config.securityKey,
      "organization_id":  config.organization_Id,
	}
	
	let request_data = {...info, ...commonParams};
	CoCreate.socket.send('createDocument', request_data, '')
}

CoCreate.socket.listen('updateDocument', function(data) {
	//. process after receive update document result
	console.log('updatedDocument', data)
})

CoCreate.socket.listen('createDocument', function(data) {
	//. createDocument result
	console.log('-----------------------------------------')
	console.log('createDocument', data)
})
